File: golang/structural/adapter/adapter.go
/*!
* \brief. Adapter. Convert the interface of a class into
*      another interface clients expect.Adapter lets classes
*      work together that couldn't otherwise because of
*      incompatible interfaces.
 */

package main

import "fmt"

// The target interface.
type Target interface {
	Request()
}

// The interface that needs to be adapted.
type IAdaptee interface {
	SpecificRequest()
}

// Adaptee.
type Adaptee struct{}

func (*Adaptee) SpecificRequest() {
	fmt.Printf("Adaptee::SpecificRequest().\n")
}

// Adapter.
type ObjectAdapter struct {
	adaptee IAdaptee
}

func NewObjectAdapter(adaptee IAdaptee) Target {
	return &ObjectAdapter{
		adaptee: adaptee,
	}
}

func (oa *ObjectAdapter) Request() {
	oa.adaptee.SpecificRequest()
}

// Adapter v2.
type ObjectAdapter2 struct {
	IAdaptee
}

func (oa *ObjectAdapter2) Request() {
	oa.IAdaptee.SpecificRequest()
}

func main() {
	// Only function SpecificRequest() can be called,
	// but all you need is function Request().
	adaptee := &Adaptee{}
	adaptee.SpecificRequest()

	// You call Request() now.
	adapter := NewObjectAdapter(adaptee)
	adapter.Request()

	// v2
	adapter2 := &ObjectAdapter{adaptee}
	adapter2.Request()
}

File: design-patterns/cpp/structural/adapter.cpp
/*!
* \brief. Adapter. Convert the interface of a class into
*      another interface clients expect.Adapter lets classes 
*      work together that couldn't otherwise because of 
*      incompatible interfaces.
*/
#include <iostream>

// The class for target interface that the customer needs.
class Target {
public:
  // Standard interface.
  virtual void Request() {
    std::cout << "Target::Request()" << std::endl;
  }
};

// The Class that need to be adapted.
class Adaptee {
public:
  void SpecificRequest() {
    std::cout << "Adaptee::SpecificRequest()" << std::endl;
  }
};

// Class Adapter. Adaptive through private inheritance.
class ClassAdapter : public Target, private Adaptee {
public:
  virtual void Request() {
    this->SpecificRequest();
  }
};

// Object Adapter. Adaptive through an object of Adaptee.
class ObjectAdapter :public Target {
public:
  ObjectAdapter(Adaptee* adaptee) {
    this->adaptee_ = adaptee_;
  }
  virtual void Request() {
    this->adaptee_->SpecificRequest();
  }
private:
  Adaptee* adaptee_;
};

int main() {
  { // Without using adapter.
    // You can't use the target interface to call the function. 
    Adaptee *src = new Adaptee();
    src->SpecificRequest();
    delete src;
  }

  std::cout << "///" << std::endl;

  { // Use Class Adapter.
    Target* target = new ClassAdapter();
    target->Request();
    delete target;
  }

  std::cout << "///" << std::endl;

  { // Use Object Adapter.
    Adaptee* src = new Adaptee();
    Target* target = new ObjectAdapter(src);
    target->Request();
    
    delete src;
    delete target;
  }

  std::cout << "///" << std::endl;

  return 0;
}